### PR TITLE
feat(tier4_autoware_api_launch): add rosbridge

### PR DESCRIPTION
**Pull request by shmpwk**
Signed-off-by: Shumpei Wakabayashi <shumpei.wakabayashi@tier4.jp>

## Description
 [web_controller](http://localhost:8085/web_controller/) needs rosbridge server.
Previously, autoware universe does not launch rosbridge because of https://github.com/autowarefoundation/autoware.universe/pull/658#issuecomment-1096445703.
So I add launching rosbridge following original  [tier4/autoware_launch](https://github.com/tier4/autoware_launch/blob/b2c2528d6c5a8d5a43b90ea9c493151ec9453853/autoware_api_launch/launch/autoware_api.launch.xml#L24-L28).

Related discussion: https://github.com/autowarefoundation/autoware/discussions/208

Note that [#318](https://github.com/autowarefoundation/autoware.universe/pull/318) removes rosbridge for fear of conflicting rosbridge in [#175](https://github.com/tier4/autoware_launch/pull/175) as they mentioned.
However  [#175](https://github.com/tier4/autoware_launch/pull/175) is under tier4 repos, but not under autowarefoundation.
Under tier4 repos, rosbridge is launched in [autoware_api_launch](https://github.com/tier4/autoware_launch/blob/b2c2528d6c5a8d5a43b90ea9c493151ec9453853/autoware_api_launch/launch/autoware_api.launch.xml#L24-L28), which cause conflict if other code lanches rosbridge.
On the other hand, under autoware foundation repos, neither [tier4_autoware_api_launch](https://github.com/autowarefoundation/autoware.universe/blob/2af223e7745509cbf7fd3c485a999f506b27be2e/launch/tier4_autoware_api_launch/launch/autoware_api.launch.xml) (same as autoware_api_launch) nor [web_controller](https://github.com/autowarefoundation/autoware.universe/blob/2af223e7745509cbf7fd3c485a999f506b27be2e/common/web_controller/launch/web_controller.launch.xml) does not launch rosbridge for fear of duplication.
See https://github.com/autowarefoundation/autoware.universe/pull/658#issuecomment-1096445703 and https://github.com/autowarefoundation/autoware.universe/pull/318.

Thus we can add launching rosbridge in autoware_api or web_controller under autoware foundation repos.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by Sharrrrk** [comment]
@shmpwk, in this case, I would propose to launch rosbridge in `web_controller.launch.xml` as the rosbridge is used by web_controller, launch it in autoware_api makes it confusing.

**Comment by Sharrrrk** [comment]
What's more, as mentioned in https://github.com/ros2/launch/issues/568 ,respawn fix has already been released into the launch packages in version 0.22.0.

**Comment by kosuke55** [comment]
FYI: https://github.com/ros2/launch/pull/569 was released in [0.22.0](https://github.com/ros2/launch/commits/0.22.0) but not in galactic.

**Comment by shmpwk** [comment]
@kosuke55 @Sharrrrk 
Thank you for the infomation.
As for `rospawn`, leave `<arg name="respawn" value="$(var respawn)"/>` as it is.

As for in which pkg should we launch rosbridge, I asked [here](https://github.com/tier4/autoware_launch/pull/175#issuecomment-1108093834)

**Comment by shmpwk** [comment]
@Sharrrrk 
Throughout discussion [here](https://github.com/tier4/autoware_launch/pull/175#issuecomment-1108118739), 
we decided to launch rosbridge in tier4_autoware_api_launch. 
Also I added docs in web_controller about rosbridge.

@yabuta 
Can you approve this PR?

**Comment by Sharrrrk** [comment]
> @Sharrrrk Throughout discussion [here](https://github.com/tier4/autoware_launch/pull/175#issuecomment-1108118739), we decided to launch rosbridge in tier4_autoware_api_launch. Also I added docs in web_controller about rosbridge.

Thank you for the detailed explanation, may be we should consider add it directly in autoware.launch.  The main reason is that web_controller is now launched by autoware.launch in core and it's somehow a compulsory part to use autoware; while for API_launch, it's still in universe and under discussion. I think in this way, it's easier for current users to understand and also doesn't block the integration of API.

**Comment by shmpwk** [comment]
@Sharrrrk 
Yes, I agree with you in that it is more straightforward to launch rosbridge in autoware.launch.xml.
My concern is rosbridge is for backend support and if it is written in autoware.launch (also in planning_simulator,launch, logging_simulator.launch), it reduces readability. A user don't have to know what is rosbridge. 
If rosbridge is written in autoware.launch, do you think rosbridge should stand alone?

**Comment by Sharrrrk** [comment]
> If rosbridge is written in autoware.launch, do you think rosbridge should be launched independently not in the module?

Yes, as far as rosbridge is a common requried component, I think it's better to be added to autoware.launch, in the tools section. Maybe with some comment like this to explain why:
```
<!-- This component allows webpages to talk with ROS, which is required by Web Controller and External API component-->
```

**Comment by shmpwk** [comment]
That seems to be a good way if rosbridge is launched with autoware_launch.
Still, I think it would be easier to use api_launch to launch rosbridge.
How about the concept that api_launch is an interface including rosbridge and web_controller is its use case?

**Comment by shmpwk** [comment]
@Sharrrrk That seems to be a good way if rosbridge is launched with autoware_launch.
Still, I think it would be easier to use api_launch to launch rosbridge.
How about the concept that api_launch is an interface including rosbridge and web_controller is its use case?

**Comment by shmpwk** [comment]
@Sharrrrk 
Looking at tier4 repos (not autowarefoundation repos), currently we put rosbridge in autoware_api.launch.
https://github.com/tier4/autoware_launch/blob/awf-latest/autoware_api_launch/launch/autoware_api.launch.xml

So my suggestion is 
1. following it as the current PR is 
2. if the PR have problem, make another PR.

File: launch/tier4_autoware_api_launch/launch/autoware_api.launch.xml [hunk 1]
```diff
@@ -1,4 +1,6 @@
 <launch>
+  <arg name="respawn" default="true"/>
```
Comment (kenji-miyake) [inline launch/tier4_autoware_api_launch/launch/autoware_api.launch.xml:2]: I guess more descriptive is better?

```suggestion
  <arg name="respawn_rosbridge" default="true"/>
```

Comment (shmpwk) [reply launch/tier4_autoware_api_launch/launch/autoware_api.launch.xml:2]: Yes thank you!

**Review by yukkysaito — APPROVED** [review]
LGTM
pls get @kenji-miyake 's aprrove as well

File: launch/tier4_autoware_api_launch/launch/autoware_api.launch.xml [hunk 2]
```diff
@@ -11,4 +13,10 @@
     <include file="$(find-pkg-share tier4_autoware_api_launch)/launch/include/internal_api_adaptor.launch.py"/>
     <include file="$(find-pkg-share tier4_autoware_api_launch)/launch/include/internal_api_relay.launch.xml"/>
   </group>
+
+  <!-- rosbridge -->
+  <!-- TODO: respawn will work once https://github.com/ros2/launch/pull/569 is released. I'll delete this comment when it's done. -->
+  <include file="$(find-pkg-share rosbridge_server)/launch/rosbridge_websocket_launch.xml">
+    <arg name="respawn" value="$(var respawn)"/>
```
Comment (kenji-miyake) [inline launch/tier4_autoware_api_launch/launch/autoware_api.launch.xml:20]:
```suggestion
    <arg name="respawn" value="$(var respawn_rosbridge)"/>
```

Comment (shmpwk) [reply launch/tier4_autoware_api_launch/launch/autoware_api.launch.xml:20]: Thank you, I haven't launched it and confirmed it works now.

**Review by kenji-miyake — APPROVED** [review]
cc @isamu-takagi

**Pull request merged**